src/libraries/System.Net.WebProxy/src/System/Net/WebProxy.cs

```diff
@@ -139,9 +138,14 @@ private void UpdateRegexList()
                         regexBypassList[i] = new Regex((string)bypassList[i]!, RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
                     }
                 }
-            }
 
-            _regexBypassList = regexBypassList;
+                _regexBypassList = regexBypassList;
+                bypassList.IsChanged = false;
```

Comment:
I believe this should be using a `Volatile.Write` to prevent reordering of these writes.

Reply:
TBH, I don't think read order will going to be changed, since compiler should know about `_regexBypassList` will be used in `UpdateRegexList()` function, if this order somehow changes, it means `code behavior in single thread` will change too. Thoughts?

Reply:
The sort of transformation I had in mind was something like
```c#
Regex[] regexBypassList = _regexBypassList;
if (_bypassList is ChangeTrackingArrayList bypassList && bypassList.IsChanged)
{
    UpdateRegexList();
    regexBypassList = _regexBypassList; // It could have changed - read it again
}

if (regexBypassList is Regex[] regexBypassList)
```

Again, I have no idea if this would ever happen in practice, but I could imagine a CPU/PGO optimization that sees the `IsChanged` branch is almost never taken and decides to prefetch `_regexBypassList`.
This doesn't impact behavior for a single thread but is a race condition if multiple threads are involved.

From our code's perspective, I think it's okay to be more defensive in such cases (e.g. making the field `volatile` instead of only doing a `Volatile.Write` on it).

Reply:
> From our code's perspective, I think it's okay to be more defensive in such cases (e.g. making the field `volatile` instead of only doing a `Volatile.Write` on it).

Well, I agree on that, so I'll do necessary changes.